syntax = "proto3";

package v1;

import "google/api/annotations.proto";
import "google/protobuf/timestamp.proto";

option go_package = "starter-kit-grpc-golang/api/gen/v1;v1";

service UserService {
  // Create User (Admin only)
  rpc CreateUser(CreateUserRequest) returns (UserResponse) {
    option (google.api.http) = {
      post: "/v1/users"
      body: "*"
    };
  }

  // Get User (Admin or Self)
  rpc GetUser(GetUserRequest) returns (UserResponse) {
    option (google.api.http) = {
      get: "/v1/users/{id}"
    };
  }

  // List Users (Admin only - with pagination/search)
  rpc ListUsers(ListUsersRequest) returns (ListUsersResponse) {
    option (google.api.http) = {
      get: "/v1/users"
    };
  }

  // Update User (Admin only - strict CRUD)
  rpc UpdateUser(UpdateUserRequest) returns (UserResponse) {
    option (google.api.http) = {
      patch: "/v1/users/{id}"
      body: "*"
    };
  }

  // Delete User (Admin only)
  rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse) {
    option (google.api.http) = {
      delete: "/v1/users/{id}"
    };
  }
}

// --- Messages ---

message UserResponse {
  string id = 1;
  string name = 2;
  string email = 3;
  string role = 4;
  bool is_email_verified = 5;
  google.protobuf.Timestamp created_at = 6;
  google.protobuf.Timestamp updated_at = 7;
}

message CreateUserRequest {
  string name = 1;
  string email = 2;
  string password = 3;
  string role = 4; // "user" or "admin"
}

message GetUserRequest {
  string id = 1;
}

message ListUsersRequest {
  int32 page = 1;
  int32 limit = 2;
  string sort = 3;   // e.g. "created_at:desc"
  string search = 4; // Search keyword
  string role = 5;   // Filter by role
  string scope = 6;  // Search scope: "name", "email", "id", or "all"
}

message ListUsersResponse {
  repeated UserResponse results = 1;
  int32 page = 2;
  int32 limit = 3;
  int32 total_pages = 4;
  int64 total_results = 5;
}

message UpdateUserRequest {
  string id = 1; // From URL
  string name = 2;
  string email = 3;
  string password = 4;
}

message DeleteUserRequest {
  string id = 1;
}

message DeleteUserResponse {
  bool success = 1;
}